### PR TITLE
FIX: Call rustdoc test with the correct cfg flags of a package.

### LINKED ISSUE #3172: Bad call to `rustdoc` when `cargo test --package feature` when using features
TLDR: this (almost) minimal project doesn't work: https://github.com/TeXitoi/cargo_bug_feature_package

When using features, and calling `cargo test --package feature`, `--cfg` arguments are bad when calling `rustdoc`.

### PR DESCRIPTION
**Pull request by jhbabon**
There was a situation in which if you you had a lib that depends on a package with features, whenever you ran the tests for the package the `rustdoc test` call was failing because it was called with the root `cfg` flags, not the package `cfg` flags.

This fix solves the issue by keeping track of the `cfg` flags per package, so the `rustdoc` command will be generated with the correct `cfg` flags.

Closes #3172

### CODE REVIEW

**Comment by TeXitoi** [comment]
Great! Thanks.

File: src/cargo/ops/cargo_rustc/mod.rs [hunk 1]
```diff
@@ -142,19 +141,21 @@ pub fn compile_targets<'a, 'cfg: 'a>(ws: &Workspace<'cfg>,
...
-    if let Some(feats) = cx.resolve.features(root_pkg) {
-        cx.compilation.cfgs.extend(feats.iter().map(|feat| {
-            format!("feature=\"{}\"", feat)
-        }));
+        if let Some(feats) = cx.resolve.features(&unit.pkg.package_id()) {
+            for feat in feats.iter() {
+                cx.compilation.cfgs.entry(unit.pkg.package_id().clone())
+                    .or_insert(HashSet::new())
+                    .insert(format!("feature=\"{}\"", feat));
+            }
```
Comment (TeXitoi) [inline src/cargo/ops/cargo_rustc/mod.rs]: You can call entry just once. Something like that should work.

```rust
cx.compilation.cfgs.entry(unit.pkg.package_id().clone())
    .or_insert(HashSet::new())
    .extend(feats.iter().map(|feat| format!("feature=\"{}\"", feat));
```

Comment (jhbabon) [inline src/cargo/ops/cargo_rustc/mod.rs]: @TeXitoi that looks better, I'll check it out.

Comment (jhbabon) [inline src/cargo/ops/cargo_rustc/mod.rs]: @TeXitoi change done here https://github.com/rust-lang/cargo/pull/3242/commits/fd8274a42407b6f1dc41674a0a9c51c7d6f6cdfe

**Comment by jhbabon** [comment]
I really don't know why the tests in travis are failing. It doesn't seem related to my changes, or it is?

Also, how can I make the tests better so the pass in appveyor? Looks like the windows version is setting extra `"` chars and the string I put in my test is not valid.

File: tests/test.rs [hunk 2]
```diff
@@ -2262,3 +2262,73 @@ fn panic_abort_multiple() {
...
+            }
+        "#);
+
+    assert_that(p.cargo_process("test")
+                .arg("--package").arg("a")
+                .arg("--verbose"),
+                execs().with_status(0)
+                       .with_stderr_contains("\
+[DOCTEST] a
+[RUNNING] `rustdoc --test [..]--cfg feature=\\\"serde_codegen\\\"[..]`"));
```
Comment (alexcrichton) [inline tests/test.rs]: Ah yeah the escaping here is subtly different on Windows (different shell), so it's fine to just match the word `serde_codegen` being passed here

**Comment by alexcrichton** [comment]
Oh it looks like Travis is failing when trying to render https://github.com/rust-lang/cargo/blob/master/src/doc/machine-readable-output.md. Could you actually modify that file as well to change the code blocks to be tagged with the `text` language instead of no language at all?

The code here looks great, thanks @jhbabon!

File: tests/test.rs (hunk 2 above)
Comment (jhbabon) [inline tests/test.rs]: @alexcrichton sounds good. I'll change it.

**Comment by jhbabon** [comment]
@alexcrichton changes done!

**Comment by alexcrichton** [comment]
@bors: r+

Thanks @jhbabon!

**Pull request merged**